stackslib/src/chainstate/nakamoto/miner.rs

```diff
@@ -833,3 +835,252 @@ impl BlockBuilder for NakamotoBlockBuilder {
         result
     }
 }
+
+/// This enum is used to supply a `reason_code` for validation
+///  rejection responses. This is serialized as an enum with string
+///  type (in jsonschema terminology).
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub enum ValidateRejectCode {
+    BadBlockHash,
+    BadTransaction,
+    InvalidBlock,
+    ChainstateError,
+    UnknownParent,
+}
+
+/// A response for block proposal validation
+///  that the stacks-node thinks should be rejected.
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct BlockValidateReject {
+    pub reason: String,
+    pub reason_code: ValidateRejectCode,
+}
+
+/// A response for block proposal validation
+///  that the stacks-node thinks is acceptable.
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct BlockValidateOk {
+    pub block: NakamotoBlock,
+    pub cost: ExecutionCost,
+    pub size: u64,
+}
+
+/// This enum is used for serializing the response to block
+/// proposal validation.
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+#[serde(tag = "Result")]
+pub enum BlockValidateResponse {
+    Ok(BlockValidateOk),
+    Reject(BlockValidateReject),
+}
+
+impl From<Result<BlockValidateOk, BlockValidateReject>> for BlockValidateResponse {
+    fn from(value: Result<BlockValidateOk, BlockValidateReject>) -> Self {
+        match value {
+            Ok(o) => BlockValidateResponse::Ok(o),
+            Err(e) => BlockValidateResponse::Reject(e),
+        }
+    }
+}
+
+impl From<Error> for BlockValidateReject {
+    fn from(value: Error) -> Self {
+        BlockValidateReject {
+            reason: format!("Chainstate Error: {value}"),
+            reason_code: ValidateRejectCode::ChainstateError,
+        }
+    }
+}
+
+impl From<DBError> for BlockValidateReject {
+    fn from(value: DBError) -> Self {
+        Error::from(value).into()
+    }
+}
+
+/// Represents a block proposed to the `v2/block_proposal` endpoint for validation
```

Comment:
Is there a reason this isn't JSON?  Seems like that would be a lot friendlier for 3rd party mining software

Reply:
Not saying you have to change it; just curious why only the `StacksMessageCodec`-specific representation is valid.

Reply:
It derives Serde `Serialize` and `Deserialize` for JSON as well. The see the integration tests, they submit `NakamotoBlockProposal`s as JSON.